src/core/Core_match.ml

```diff
@@ -215,22 +215,23 @@ end)
 
 (* Deduplicate matches *)
 let uniq (pms : t list) : t list =
-  (* perf: Previously used an initial size of 1024 but profiling with memtrace
-      revealed that this was oversized and contributed to 14% of all major heap
-      usage in an interfile run over juice-shop. Setting it to 8 meant it was
-      undersized in some cases and led to additional allocations as the table
-      gets resized. Change this only with some caution.
-      DOUBT(iago): What is the actual observed effect in terms of runtime and
-        peak memory usage?
-  *)
-  let size = List.length pms in
-  let matches_tbl = Tbl.create size in
-  (* dedup matches *)
-  pms
-  |> List.iter (fun match_ ->
-         if not (Tbl.mem matches_tbl match_) then Tbl.add matches_tbl match_ ());
-  (* list matches *)
-  Tbl.fold (fun m () acc -> m :: acc) matches_tbl []
+  let digest (t : t) =
+    (t.rule_id.id,
+     t.path,
+     (fst t.range_loc).pos.bytepos,
+     (snd t.range_loc).pos.bytepos,
+     t.env)
+  in
+  let cmp t1 t2 =
+    let c = compare (digest t1) (digest t2) in
+    if c <> 0 then c else
+      match t1.taint_trace, t2.taint_trace with
+      | None, None -> 0
+      | None, _ -> 1
+      | Some _, None -> -1
```

Comment:
`_, None` would also work, since if the lhs was `None` it would have been caught by the case above.